workflow-ui/core-common/src/main/java/com/squareup/workflow1/visual/WithEnvironment.kt

```diff
@@ -0,0 +1,33 @@
+package com.squareup.workflow1.visual
+
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+
+@WorkflowUiExperimentalApi
+public abstract class WithEnvironment<W : Any>(
+  wrapped: W,
+  public val environment: VisualEnvironment
+) : Wrapper<W>(wrapped)
+
+@WorkflowUiExperimentalApi
+public class WithEnvironmentVisualFactory<C, W : Any, V> : VisualFactory<C, WithEnvironment<W>, V> {
+  override fun createOrNull(
+    rendering: WithEnvironment<W>,
+    context: C,
+    environment: VisualEnvironment,
+    getFactory: (VisualEnvironment) -> VisualFactory<C, Any, V>
+  ): VisualHolder<WithEnvironment<W>, V> {
+    val mergedEnvironment = environment + rendering.environment
+
+    val delegateHolder = getFactory(mergedEnvironment).create(
+      rendering.wrapped, context, mergedEnvironment, getFactory
+    )
```

Comment:
@helios175 This is why `getFactory` needs the `VisualEnvironment` argument -- so that `VisualFactory` implementations can mess with it. That's a primary use case.